Connectors/Rhino/Speckle.Connectors.GrasshopperShared/Operations/Send/GrasshopperBlockPacker.cs

```diff
@@ -22,42 +26,80 @@ internal sealed class GrasshopperBlockPacker
+  public List<SpeckleObjectWrapper>? ProcessInstance(SpeckleBlockInstanceWrapper? blockInstance, int depth = 0)
   {
-    // NOTE: in pure gh workflows Definition might be null ALTHOUGH we're throwing an Exception 🤨. But rather safe than sorry.
-    if (blockInstance?.Definition != null)
+    if (blockInstance?.Definition == null)
     {
-      return ProcessDefinition(blockInstance.Definition);
+      return null;
     }
 
-    return null;
+    blockInstance.ApplicationId ??= Guid.NewGuid().ToString();
+    var instanceId = blockInstance.ApplicationId;
+
+    blockInstance.InstanceProxy.maxDepth = depth;
+    _instanceObjectsManager.AddInstanceProxy(instanceId, blockInstance.InstanceProxy);
+
+    return ProcessDefinition(blockInstance.Definition, depth);
   }
 
   /// <summary>
-  /// Processes a <see cref="SpeckleBlockDefinitionWrapper"/> and adds it to the current collection (if not already present).
+  /// Processes a block definition, adding it and its objects to InstanceObjectsManager.
+  /// Updates maxDepth for existing definitions when encountered at greater depths.
   /// </summary>
-  private List<SpeckleObjectWrapper>? ProcessDefinition(SpeckleBlockDefinitionWrapper definition)
+  private List<SpeckleObjectWrapper>? ProcessDefinition(SpeckleBlockDefinitionWrapper definition, int depth = 0)
   {
     // Use wrapper's id as definitive identifier. Create if empty.
     definition.ApplicationId ??= Guid.NewGuid().ToString();
     string definitionId = definition.ApplicationId;
 
-    // Only add if not in InstanceDefinitionProxies
-    if (!InstanceDefinitionProxies.ContainsKey(definitionId))
+    // Check if already processed using InstanceObjectsManager
+    if (
+      _instanceObjectsManager.TryGetInstanceDefinitionProxy(definitionId, out InstanceDefinitionProxy? definitionProxy)
+    )
     {
-      // 💩 Sync proxy appId to wrapper appId. Can this mismatch even occur? Is this the right approach?
-      InstanceDefinitionProxy proxy = definition.InstanceDefinitionProxy;
-      proxy.applicationId = definitionId;
-      InstanceDefinitionProxies[definitionId] = proxy;
+      int depthDifference = depth - definitionProxy.maxDepth;
+      if (depthDifference > 0)
+      {
+        // Use InstanceObjectsManager to update max depth
+        _instanceObjectsManager.UpdateChildrenMaxDepth(definitionProxy, depthDifference);
+      }
 
-      // Return objects to be added to collection (only happens once per definition)
-      return definition.Objects;
+      return null; // this prevents infinite recursion
     }
 
-    // Already processed this definition - don't return objects
-    return null;
+    // Process objects recursively
+    var objectsToAdd = new List<SpeckleObjectWrapper>();
+    foreach (var obj in definition.Objects)
+    {
+      if (obj is SpeckleBlockInstanceWrapper nestedInstance)
+      {
+        objectsToAdd.Add(nestedInstance);
```

Comment:
lines 81-82 and 93-94 can be moved up to the foreach loop (79)

If the nested instance doesn't have an id at this point, then prbly should add a warning here and skip instead of assigning? because it shouldn't happen right?

Reply:
agreed on both points. See [refactor: consolidate object processing in GrasshopperBlockPacker](https://github.com/specklesystems/speckle-sharp-connectors/pull/934/commits/b3259a7330abea650718116ef495f950a95aa485)